### PR TITLE
Reopen Optimized deep delete #170

### PR DESCRIPTION
**Pull request by johannes-vogel**
This still causes some runtime tests to fail. We need to analyze before merging as this causes the runtime PRs to be blocked.

- [x] requires https://github.com/cap-js/cds-dbs/pull/213
- [x] requires https://github.com/cap-js/cds-dbs/pull/214 to be fixed
- [x] revert to hierarchy support and throw error for cycles
- [x] adapt runtime tests with cyclic compositions

### CODE REVIEW

File: db-service/lib/SQLService.js [hunk 1]
```diff
@@ -168,13 +168,21 @@ class SQLService extends DatabaseService {
...
-        await Promise.all (Object.values(compositions).map(c => {
+        for(const c of Object.values(compositions)) {
           if (c._target['@cds.persistence.skip'] === true) return
-          if (c._target === req.target) if (++depth > (c['@depth'] || 3)) return
-          let query = DELETE.from ({ref:[ ...from.ref, c.name ]}) // CQL`DELETE from Foo[pred]:comp1.comp2...`
-          return this.onDELETE({ query, depth, target: c._target })
-        }))
+          let count = 0
+          for (let i = 0; i < depth.length; i++) {
+            if (depth[i] === c) count++
```
Comment (johannes-vogel) [inline db-service/lib/SQLService.js:177]: revert to hierarchies and throw in case of circle detected.

File: db-service/lib/SQLService.js [hunk 2]
```diff
@@ -380,5 +410,10 @@ cds.extend(cds.ql.Query).with(
   },
 )
 
+const _circular_composition = (visited, target) => new Error(`
+  Transitive circular composition detected:
+  ${visited.join(' > ')} > ${target}
+  These are not supported by deep delete.`
+)
```
Comment (danjoa) [inline db-service/lib/SQLService.js:417]:
```suggestion
```

**Review by danjoa — APPROVED** [review]

File: db-service/lib/SQLService.js (hunk 1 above)
Comment (danjoa) [reply db-service/lib/SQLService.js]: Sollte jetzt passen

File: db-service/lib/SQLService.js [hunk 3]
```diff
@@ -152,6 +149,39 @@ class SQLService extends DatabaseService {
...
+        let { from, where } = req.query.DELETE
+        if (typeof from === 'string') from = { ref: [from] }
+        if (where) {
+          let last = from.ref.at(-1)
+          if (last.where) [ last, where ] = [ last.id, [ { xpr: last.where }, 'and', { xpr: where } ] ]
+          from = {ref:[ ...from.ref.slice(0,-1), { id: last, where }]}
+        }
+        // Process child compositions depth-first
+        let { depth=0, visited=[] } = req
+        visited.push (req.target.name)
```
Comment (johannes-vogel) [inline db-service/lib/SQLService.js:167]: Do we support sth. like 
```cds
entity  E {
  ...
  c1: Composition of X...
  c2: Composition of X...
}
```
the current implementation does not allow parallel compositions to same target.

**Comment by johannes-vogel** [comment]
I'm merging with admin power now. Seems something is off with the codeowner file.

**Pull request merged**